File: JavaFactorial/src/de/luschny/math/Xmath.java
// Copyright (C) 2004-2009 Peter Luschny, MIT License applies.
// See http://en.wikipedia.org/wiki/MIT_License
// Visit http://www.luschny.de/math/factorial/FastFactorialFunctions.htm
// Comments mail to: peter(at)luschny.de
package de.luschny.math;

import de.luschny.math.arithmetic.Xint;

/**
 * Class declaration
 * 
 * @author
 * @version %I%, %G%
 */
public final class Xmath {

    /**
     * Method declaration
     * 
     * @param w
     * @return
     */
    public static int bitCount(int w) {
        w -= (0xaaaaaaaa & w) >>> 1;
        w = (w & 0x33333333) + ((w >>> 2) & 0x33333333);
        w = w + (w >>> 4) & 0x0f0f0f0f;
        w += w >>> 8;
        w += w >>> 16;

        return w & 0xff;
    }

    /**
     * Method declaration
     * 
     * @param w
     * @return
     */
    // Binaeres Suchen - Entscheidungsbaum (5 Tests, selten 6)
    public static int bitLen(int w) {
        return w < 1 << 15 ? (w < 1 << 7 ? (w < 1 << 3 ? (w < 1 << 1 ? (w < 1 << 0 ? (w < 0 ? 32 : 0) : 1)
                : (w < 1 << 2 ? 2 : 3)) : (w < 1 << 5 ? (w < 1 << 4 ? 4 : 5) : (w < 1 << 6 ? 6 : 7)))
                : (w < 1 << 11 ? (w < 1 << 9 ? (w < 1 << 8 ? 8 : 9) : (w < 1 << 10 ? 10 : 11))
                : (w < 1 << 13 ? (w < 1 << 12 ? 12 : 13) : (w < 1 << 14 ? 14 : 15))))
                : (w < 1 << 23 ? (w < 1 << 19 ? (w < 1 << 17 ? (w < 1 << 16 ? 16 : 17) : (w < 1 << 18 ? 18 : 19))
                : (w < 1 << 21 ? (w < 1 << 20 ? 20 : 21) : (w < 1 << 22 ? 22 : 23)))
                : (w < 1 << 27 ? (w < 1 << 25 ? (w < 1 << 24 ? 24 : 25) : (w < 1 << 26 ? 26 : 27))
                : (w < 1 << 29 ? (w < 1 << 28 ? 28 : 29) : (w < 1 << 30 ? 30 : 31))));
    }
    /**
     * Method declaration
     * 
     * @param n
     * @return
     * @throws IllegalArgumentException
     *             <tt>20 &lt; n</tt>.
     */
    private static long[] smallFactorials = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880, 3628800, 39916800, 479001600,
        6227020800L, 87178291200L, 1307674368000L, 20922789888000L, 355687428096000L,
        6402373705728000L, 121645100408832000L, 2432902008176640000L};

    public static Xint Factorial(int n) {
        if (n > 19) {
            throw new IllegalArgumentException("n < 20 required");
        }
        if (n < 0) {
            throw new IllegalArgumentException("0 <= n required");
        }

        return Xint.valueOf(smallFactorials[n]);
    }

    /**
     * status :: hack
     * 
     * @param n
     * @return
     */
    public static int piHighBound(long n) {
        if (n < 17) {
            return 6;
        }
        return (int) Math.floor(n / (Math.log(n) - 1.5));
    }

    /**
     * @param n
     * @return
     */
    public static long naiveSwingFactorial(int n) {
        if (n < 1) {
            return 1;
        }

        if ((n & 1) == 1) {
            return naiveSwingFactorial(n - 1) * n;
        }

        return (naiveSwingFactorial(n - 1) << 2) / n;
    }

    /**
     * @param value
     * @return <tt>log<sub>10</sub>value</tt>.
     */
    public static double log10(double value) {
        return Math.log(value) * 0.43429448190325176;
    }

    /**
     * @param value
     * @return <tt>log<sub>2</sub>value</tt>.
     */
    public static double log2(double value) {
        return Math.log(value) * 1.4426950408889634;
    }

    /**
     * @param n
     * @return (int) floor(log[2](n))
     */
    public static int floorLog2(int n) {
        if (n <= 0) {
            throw new IllegalArgumentException("n > 0 required");
        }
        return bitLen(n) - 1;
    }

    /**
     * @param n
     * @return (int) ceil(log[2](n))
     */
    public static int ceilLog2(int n) {
        int ret = floorLog2(n);
        if (n != (1 << ret)) {
            ret++;
        }
        return ret;
    }

    /**
     * Berechnung der groessten Zahl, deren Quadrat n nicht uebersteigt.
     * 
     * @param n
     * @return (int) floor(sqrt(n))
     */
    public static int sqrt(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("arg >= 0 required");
        }
        if (n == 0) {
            return 0;
        }

        int unten, oben = bitLen(n);

        do {

            unten = n / oben;
            oben += unten;
            oben >>= 1;
        } while (oben > unten);

        return oben;
    }

    /**
     * Method declaration
     * 
     * @param n
     * @return
     */
    public static int omegaSwingHighBound(int n) {
        if (n < 4) {
            return 6;
        }

        return (int) Math.floor(Math.sqrt(n) + n / (log2(n) - 1.0));
    }

    /**
     * @param x
     * @return
     */
    public static double asymptFactorial(double x) {
        // asymptotic error of order O(x^-5)
        // double ln2Pi = 1.8378770664093455 = Math.log(2 * Math.PI);
        double a = x + x + 1;
        return (1.8378770664093455 + Math.log(a / 2) * a - a - (1 - 7 / (30 * a * a)) / (6 * a)) / 2;
    }

    /**
     * @param x
     * @return
     */
    public static String asymptFactorial(int x) {
        // error of order O(x^-5)
        return exp(asymptFactorial((double) x));
    }

    /**
     * @param x
     * @return
     */
    public static String exp(double x) {
        double l = x * 0.43429448190325176; // x/Math.log(10);
        double e = Math.floor(l);
        double m = Math.pow(10, l - e);
        String mat = (Double.toString(m)).substring(0, 6);
        return mat + " E+" + Integer.toString((int) e);
    }

    /**
     * @param x
     * @return
     */
    public static double asymptSwingingFactorial(double x) {
        // not yet implemented
        return x * 0;
    }

    public static long gcd(long a, long b) {
        long x, y;

        if (a < 0) {
            a = -a;
        }
        if (b < 0) {
            b = -b;
        }

        if (a >= b) {
            x = a;
            y = b;
        } else {
            x = b;
            y = a;
        }

        while (y != 0) {
            long t = x % y;
            x = y;
            y = t;
        }
        return x;
    }

    public static double round(double d, int scale, int mode) {
        java.math.BigDecimal bd = new java.math.BigDecimal(Double.toString(d));
        return (bd.setScale(scale, mode)).doubleValue();
    }

    public static double round2(double d) {
        return round(d, 2, java.math.BigDecimal.ROUND_HALF_UP);
    }

    /**
     * We don't need a constructor.
     */
    private Xmath() {
    }
}
// endOfXmath

File: SilverFactorial64/Sharith/MathUtils/XMath.cs
// -------- ToujoursEnBeta
// Author & Copyright : Peter Luschny
// License: LGPL version 3.0 or (at your option)
// Creative Commons Attribution-ShareAlike 3.0
// Comments mail to: peter(at)luschny.de
// Created: 2010-03-01

namespace Sharith.MathUtils
{
    using System;
    using System.Threading.Tasks;

    using XInt = Sharith.Arithmetic.XInt;

    public static class XMath
    {
        /// <summary>
        /// Bit count
        /// </summary>
        /// <param name="w"></param>
        /// <returns>Number of bits.</returns>
        public static int BitCount(int w)
        {
            w -= (int)((0xaaaaaaaa & w) >> 1);
            w = (w & 0x33333333) + ((w >> 2) & 0x33333333);
            w = w + (w >> 4) & 0x0f0f0f0f;
            w += w >> 8;
            w += w >> 16;

            return w & 0xff;
        }

        /// <summary>
        /// Calculates the bit length of an integer.
        /// </summary>
        /// <param name="w"></param>
        /// <returns>bit length</returns>
        public static int BitLength(int w)
        {
            return (w < 1 << 15
            ? (w < 1 << 7 ? (w < 1 << 3 ? (w < 1 << 1 ? (w < 1 << 0 ? (w < 0 ? 32 : 0) : 1)
            : (w < 1 << 2 ? 2 : 3)) : (w < 1 << 5 ? (w < 1 << 4 ? 4 : 5) : (w < 1 << 6 ? 6 : 7)))
            : (w < 1 << 11 ? (w < 1 << 9 ? (w < 1 << 8 ? 8 : 9) : (w < 1 << 10 ? 10 : 11))
            : (w < 1 << 13 ? (w < 1 << 12 ? 12 : 13) : (w < 1 << 14 ? 14 : 15))))
            : (w < 1 << 23
            ? (w < 1 << 19 ? (w < 1 << 17 ? (w < 1 << 16 ? 16 : 17) : (w < 1 << 18 ? 18 : 19))
            : (w < 1 << 21 ? (w < 1 << 20 ? 20 : 21) : (w < 1 << 22 ? 22 : 23)))
            : (w < 1 << 27
            ? (w < 1 << 25 ? (w < 1 << 24 ? 24 : 25) : (w < 1 << 26 ? 26 : 27))
            : (w < 1 << 29 ? (w < 1 << 28 ? 28 : 29)
            : (w < 1 << 30 ? 30 : 31)))));
        }

        public static int BitLength(byte b)
        {
            return b < 8 ? b < 2 ? b < 1 ? 0 : 1 : b < 4 ? 2 : 3 : b < 32
                 ? b < 16 ? 4 : 5 : b < 64 ? 6 : 7;
        }

        //public static int BitLength(XInt n)
        //{
        //    byte[] bytes = (n.Sign * n).ToByteArray();
        //    int i = bytes.Length - 1;
        //    return (i << 3) + BitLength(bytes[i]);
        //}

        /// <summary>
        /// High bound for number of primes not exeeding n.
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int PiHighBound(long n)
        {
            if (n < 17)
            {
                return 6;
            }
            return (int)System.Math.Floor(((double)n)
                                          / (System.Math.Log(n) - 1.5));
        }

        /// <summary>
        /// Logarithm to base 10.
        /// </summary>
        /// <param name="value"></param>
        /// <returns><tt>Log<sub>10</sub>value</tt>.</returns>
        //public static double Log10(double value)
        //{
        //    return System.Math.Log(value) * 0.43429448190325176;
        //}

        /// <summary>
        /// Logarithm to base 2.
        /// </summary>
        /// <param name="value"></param>
        /// <returns><tt>Log<sub>2</sub>value</tt>.</returns>
        public static double Log2(double value)
        {
            return System.Math.Log(value) * 1.4426950408889634;
        }

        /// <summary>
        /// Floor of the binary logarithm.
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int FloorLog2(int n)
        {
            if (n <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(n) + " >= 0 required");
            }
            return BitLength(n) - 1;
        }

        /// <summary>
        /// Floor of the square root.
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int FloorSqrt(int n)
        {
            if (n < 0)
            {
                throw new System.ArgumentOutOfRangeException(nameof(n) + " >= 0 required");
            }
            return (int)System.Math.Floor(System.Math.Sqrt(n));
        }

        /// <summary>
        /// Ceiling of the binary logarithm.
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int CeilLog2(int n)
        {
            int ret = FloorLog2(n);
            if (n != (1 << ret)) ret++;
            return ret;
        }

        /// <summary>
        /// Berechnung der groessten Zahl, deren Quadrat n nicht uebersteigt.
        /// </summary>
        /// <param name="n"></param>
        /// <returns></returns>
        public static int Sqrt(int n)
        {
            if (n < 0)
            {
                throw new System.ArgumentOutOfRangeException(nameof(n) + " >= 0 required");
            }

            if (n == 0) return 0;

            int unten, oben = BitLength(n);

            do
            {
                unten = n / oben;
                oben += unten;
                oben >>= 1;
            }
            while (oben > unten);

            return oben;
        }

        public static int OmegaSwingHighBound(int n)
        {
            if (n < 4)
            {
                return 6;
            }
            return (int)System.Math.Floor(System.Math.Sqrt(n)
                                          + (double)n / (Log2(n) - 1));
        }

        public static double AsymptFactorial(double x)
        {
            // error of order O(x^-5)
            // double ln2Pi = 1.8378770664093455 = Math.log(2 * Math.PI);
            double a = x + x + 1;
            return (1.8378770664093455 + System.Math.Log(a / 2) * a - a
                    - (1 - 7 / (30 * a * a)) / (6 * a)) / 2;
        }

        public static string AsymptFactorial(int x)
        {
            // error of order O(x^-5)
            return Exp(AsymptFactorial((double)x));
        }

        public static long ExactDecimalDigitsPerMillisecond(int n, long ms)
        {
            double x = AsymptFactorial((double)n);
            double l = x * 0.43429448190325176D; // x/Math.log(10);
            double lsd = ms == 0 ? l : l / (double)ms;
            return (long)Math.Truncate(0.5 + lsd);
        }

        public static string Exp(double x)
        {
            double l = x * 0.43429448190325176D; // x/Math.log(10);
            double e = System.Math.Floor(l);
            double m = System.Math.Pow(10, l - e);
            string mat = (Convert.ToString(m)).Substring(0, 6);
            return mat + " E+" + Convert.ToString((int)e);
        }

        public static double AsymptSwingingFactorial(double x)
        {
            // TODO: not yet implemented
            return 1;
        }

        public static void ClearOpCounter()
        {
            // --
        }
        public static long[] GetOpCounts()
        {
            return new long[] { 1 };
        }

        private static long[] smallFactorials = {
        1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880, 3628800, 39916800, 
        479001600, 6227020800, 87178291200, 1307674368000, 20922789888000, 
        355687428096000, 6402373705728000, 121645100408832000, 2432902008176640000 };

        public static XInt Factorial(int n)
        {
            if (n > 20)
            {
                throw new System.ArgumentOutOfRangeException("max n is 20 but was " + n.ToString());
            }
            if (n < 0)
            {
                throw new System.ArgumentOutOfRangeException("min " + nameof(n) + " is 0");
            }

            return new XInt(smallFactorials[n]);
        }

        const int ParallelThreshold = 1024;

        // <returns>a[start]*a[start+1]*...*a[start+length-1]</returns>
        public static XInt Product(int[] a, int start, int length)
        {
            if (length == 0) return XInt.One;

            int len = (length + 1) / 2;
            long[] b = new long[len];

            int i, j, k;

            for (k = 0, i = start, j = start + length - 1; i < j; i++, k++, j--)
            {
                b[k] = a[i] * (long)a[j];
            }

            if (i == j) b[k++] = a[j];

            if (k > ParallelThreshold)
            {
                var task = Task.Factory.StartNew<XInt>(() =>
                {
                    return RecProduct(b, (k - 1) / 2 + 1, k - 1);
                });

                var left = RecProduct(b, 0, (k - 1) / 2);
                var right = task.Result;
                return left * right;
            }

            return RecProduct(b, 0, k - 1);
        }

        public static XInt Product(int[] a, int start, int length, int increment)
        {
            if (length == 0) return XInt.One;

            int len = (1 + (length + 1) / 2) / increment;
            long[] b = new long[len];

            int i, k = 0;
            bool toggel = false;

            for (i = start; i < start+length; i += increment)
            {
                if ((toggel = !toggel)) b[k] = a[i];
                else b[k++] *= (long)a[i];
            }
       
            if (len > ParallelThreshold)
            {
                var task = Task.Factory.StartNew<XInt>(() =>
                {
                    return RecProduct(b, (len - 1) / 2 + 1, len - 1);
                });

                var left = RecProduct(b, 0, (len - 1) / 2);
                var right = task.Result;
                return left * right;
            }

            return RecProduct(b, 0, len - 1);
        }

        public static XInt Product(int[] a)
        {
            return Product(a, 0, a.Length);
        }

        public static XInt Product(long[] a, int len)
        {
            int n = len - 1;
            if (len > ParallelThreshold)
            {
                var task = Task.Factory.StartNew<XInt>(() =>
                {
                    return RecProduct(a, n / 2 + 1, n);
                });

                var left = RecProduct(a, 0, n / 2);
                var right = task.Result;
                return left * right;
            }

            return RecProduct(a, 0, n);
        }

        public static XInt Product(long[] a)
        {
            return Product(a, a.Length);
        }

        public static XInt RecProduct(long[] s, int n, int m)
        {
            if (n > m)
            {
                return XInt.One;
            }
            if (n == m)
            {
                return new XInt(s[n]);
            }

            int k = (n + m) >> 1;
            return RecProduct(s, n, k) * RecProduct(s, k + 1, m);
        }

        private static XInt BigRecProduct(XInt[] s, int n, int m)
        {
            if (n > m)
            {
                return XInt.One;
            }
            if (n == m)
            {
                return s[n];
            }

            int k = (n + m) >> 1;
            return BigRecProduct(s, n, k) * BigRecProduct(s, k + 1, m);
        }

        public static XInt Product(XInt[] a, int start, int len)
        {
            int n = len - 1;
            if (len > ParallelThreshold)
            {
                var task = Task.Factory.StartNew<XInt>(() =>
                {
                    return BigRecProduct(a, start + n / 2 + 1, start + n);
                });

                var left = BigRecProduct(a, start, start + n / 2);
                return left * task.Result;
            }

            return BigRecProduct(a, start, n);
        }

        public static XInt Product(XInt[] a)
        {
            return BigRecProduct(a, 0, a.Length - 1);
        }

    } 
} 


//static public class MathFun
//{
//    // Calibrate the treshhold
//    private const int THRESHOLD_PRODUCT_SERIAL = 128;

//    //static public XInt ProductSerial(long[] seq, int start, int len)
//    //{
//    //    var prod = new XInt(seq[start]);
//    //    for (int i = start + 1; i < start + len; i++)
//    //    {
//    //        prod *= seq[i];
//    //    }
//    //    return prod;
//    //}

//    static public XInt Product(long[] seq, int start, int len)
//    {
//        if (len <= THRESHOLD_PRODUCT_SERIAL)
//        {
//            // return ProductSerial(seq, start, len);
//            var rprod = new XInt(seq[start]);

//            for (int i = start + 1; i < start + len; i++)
//            {
//                rprod *= seq[i];
//            }
//            return rprod;
//        }
//        else
//        {
//            int halfLen = len / 2;

//            Task<XInt> task = Task.Factory.StartNew(() => Product(seq, start, halfLen));
//            var rprod = Product(seq, start + halfLen, len - halfLen);
//            return task.Result * rprod;

//            // rprod = XInt.Zero; XInt lprod = XInt.Zero;
//            //Parallel.Invoke(
//            //    () => { rprod = Product(seq, start, halfLen); },
//            //    () => { lprod = Product(seq, start + halfLen, len - halfLen); }
//            //);
//            //rprod = lprod * rprod;
//        }
//    }

//    private MathFun() { }

// System.Diagnostics.Debug.Assert((0 <= start) & (length <= a.Length),
// System.Reflection.MethodBase.GetCurrentMethod().Name + " failed: range not subrange of a");
//}

